### PR TITLE
Introduce Create Tenant API Endpoint and "USDR Super Admin" access helper

### PR DESCRIPTION
**Pull request by slants3**
### Ticket #
[84](https://github.com/usdigitalresponse/usdr-gost/issues/84)

### Description
To add new tenants, we need to create an API endpoint.

This replicates our existing API infrastructure currently used for adding users.

### CODE REVIEW

File: packages/server/src/routes/tenants.js
Comment (Rapol) [inline packages/server/src/routes/tenants.js:29]: i might be getting this confused, since this tenant entity is new to me. A tenant is the main entity and agencies are then created under that tenant, correct? So I think we will not have agency_id.

1. Create Nevada tenant
2. Create Grant Office tenant
3. Update Nevada tenant as the main agency

Im not entirely sure what is the usage of main_agency_id

Comment (mhuang-usdr) [reply packages/server/src/routes/tenants.js:29]: the agencies are in a tree hierarchy under the tenant. `main_agency_id` is the root of the tree - the top-level agency that belongs to the tenant. This agency is required when the tenant is first created.

File: packages/server/src/routes/tenants.js
Comment (bspates) [inline packages/server/src/routes/tenants.js:32]: Should we be doing any validation/sanitization of the request data before sending it  to the DB, or is that built in somewhere else?

File: packages/server/src/routes/tenants.js
Comment (slants3) [reply packages/server/src/routes/tenants.js:29]: @mhuang-usdr I'm a little confused - if the agency belongs to the tenant, couldn't it be possible for a tenant to be created before its agency is created?

File: packages/server/src/routes/tenants.js
Comment (slants3) [reply packages/server/src/routes/tenants.js:32]: That's a great idea - as far as I understand it there aren't any sort of validation built into our endpoints as is the case here with [users](https://github.com/usdigitalresponse/usdr-gost/blob/4389f12e9aa797b77d4102337d3480491c9efa27/packages/server/src/routes/users.js#L8) as well.

Since we don't have any validation or sanitation anywhere - what do you think about adding this endpoint for now to unblock https://github.com/usdigitalresponse/usdr-gost/pull/189 and then creating a separate issue where we add validation + sanitation to all of our endpoints?

Comment (bspates) [reply packages/server/src/routes/tenants.js:32]: Makes sense to me.

File: packages/server/src/routes/tenants.js
Comment (mhuang-usdr) [reply packages/server/src/routes/tenants.js:29]: @slants3 I apologize, I realize forgot to explain this earlier!

The requirements for the app to function as expected are:
- Every tenant must have at least one agency - the main agency 
- Every agency must have at least one user, and they must be an admin.

Thus the requirements for spinning up a new tenant are:
- A new tenant in the tenant database
- A new agency assigned to the new tenant ("Grants Office" by default)
- A new admin user ("grants.dev@usdigitalresponse.org" by default) assigned to that tenant

I'm not sure what the best order or best practices to create those entries are, that's up to the eng team!


(Also, a possible snag - iirc users _might_ still be keyed by their email. In which case having the same user (grants.dev@usdigitalresponse.org) across multiple tenants will cause issues. The desired experience is that we _are_ able to have the same user in multiple tenants, but in the interim you can use "grants.dev+nv@usdigitalresponse.org" for the Nevada tenant.

Comment (slants3) [reply packages/server/src/routes/tenants.js:29]: Makes total sense, thanks for explaining that!

Comment (mhuang-usdr) [reply packages/server/src/routes/tenants.js:29]: @slants3 we should create a new agency when we create a new tenant. Going forward, we should never need to reuse an agency for a new tenant. 

That is only happening right now due to a database cleanup (there are a lot of things in there that got really messed up)

Comment (slants3) [reply packages/server/src/routes/tenants.js:29]: I see - what is the workflow for the agency + tenant creation? I'm trying to understand if they should be happening synchronously or if the new agency can be created first and then just passed into the tenant creation as an `main_agency_id`.

Comment (mhuang-usdr) [reply packages/server/src/routes/tenants.js:29]: @slants3 no current existing workflow. The agency should be created as the tenant is created.  Here's an example user journey.

Mindy, the PM, wants to onboard a new tenant. She runs the script (or, when we have a UI down the road, clicks the button), and a new tenant will be automatically set up with the main agency and default admin user (grants.dev@usdigitalresponse.org)

File: packages/server/src/routes/tenants.js
Comment (slants3) [inline packages/server/src/routes/tenants.js:35]: Hey @bspates, I had a quick question for you - I'm using the updated `createAgency` endpoint but I'm unable to successfully create an agency and the creation hangs. I was able to create before the change - but as far as I can tell I'm using this correctly. Would you have an idea on what I'm doing wrong here?

Comment (bspates) [reply packages/server/src/routes/tenants.js:35]: @slants3 Do you get any error? What is the hardcoded number 15? Shouldn't you be using the user id that is making the request?

Comment (bspates) [reply packages/server/src/routes/tenants.js:35]: @slants3 Also I see the tenant creation is happening after this. You'll want to swap those around. How the foreign key dependencies in the tables work now is that only main_agency_id in tenant can be left null to start. So you have to create the tenant, then the agency, then go back and update the main_agency_id in the tenant.

File: packages/server/src/db/index.js
Comment (mattbroussard) [inline packages/server/src/db/index.js:581]: Why not just get the whole row from Postgres and return that?

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [inline packages/server/src/routes/tenants.js:20]: This route doesn't seem to check any kind of permission. I think we might want a new helper like `requireUSDRAdminUser` that calls `requireAdminUser` and additionally verifies the user is part of a hardcoded agency ID (`1`, I think).

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [reply packages/server/src/routes/tenants.js:35]: >Shouldn't you be using the user id that is making the request?

👍, should use `req.session.user.id` (available when using `requireAdminUser` as suggested in a separate comment)

>So you have to create the tenant, then the agency, then go back and update the main_agency_id in the tenant.

👍, this should also mean `setAgencyTenantId` is not necessary since agencies must be created with a `tenant_id` to start with and cannot move between tenants.

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [inline packages/server/src/routes/tenants.js:43]: nit: remove `_ =`

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [inline packages/server/src/routes/tenants.js:55]: This comment will no longer be true; revise.

**Review by mattbroussard — CHANGES_REQUESTED** [review]
Is the idea that this will be called from some UI (e.g. added to currently-unused `Tenants.vue`?) that is gated to only USDR admins? Or that this endpoint will be called from some script we run? If the latter, why have an endpoint at all vs. having a script that just does this directly?

**Comment by mattbroussard** [comment]
Just ran across #193 -- is this intended to supersede that, or will we have _both_ UI for this and a script?

File: packages/server/src/routes/tenants.js
Comment (slants3) [reply packages/server/src/routes/tenants.js]: > Just ran across #193 -- is this intended to supersede that, or will we have _both_ UI for this and a script?

Referencing this - I think in the immediate term we're going to use a script, but in the long term we will eventually want this to be an UI element

**Comment by mattbroussard** [comment]
As [mentioned](https://usdigitalresponse.slack.com/archives/C0324KDQSCR/p1662679750039039) on Slack, taking this over because I need this logic for ARPA Integration.

In update a bit ago, I've mostly rewritten much of the PR. Notably:
- Reuses most of the logic I wrote in #193's script in the endpoint
  - I also moved the file. Since there is both a move and modifications to it, Github confusingly shows the entire file `tenant_creation.js` as new, but the bulk of the logic in it was actually implemented in #193 and has just been slightly reorganized in this PR.
- Adds a new access check middleware, `requireUSDRSuperAdminUser` based on a hardcoded tenant ID, and uses it on these routes that view/modify across tenants.
  - However, it looks like in prod there are lots of non-USDR users as "admins" with tenantId=1. @mhuang-usdr is there a different condition we should be using for "super admin", or should we clean up this prod data so only actual USDR staff/volunteers are in the USDR tenant?
- Some small tweaks to the UI of the tenants tab (gating its display (superseding #227), showing all tenants for super-admins)

I might go ahead and hack up a simple "Create Tenant" modal since it seems silly to have this endpoint but nothing that calls it?

File: packages/server/src/lib/access-helpers.js
Comment (mattbroussard) [inline packages/server/src/lib/access-helpers.js:15]: Thinking more about this: maybe the right thing to do actually is to create a third `role` instead of hard-coding tenant/agency IDs like this. Would require at least:
1. Finding places that hardcode checks for admin role to also allow super-admin (including in soon-to-be-merged ARPA Reporter code)
2. Adding a filter to what roles someone can set a user to when creating or editing a user. Only super-admin can designate others as super-admin.

@bspates @mhuang-usdr thoughts?

File: packages/server/src/db/tenant_creation.js
Comment (ajhyndman) [inline packages/server/src/db/tenant_creation.js:1]: Nit: I suspect this might be redundant with #323

Comment (mattbroussard) [reply packages/server/src/db/tenant_creation.js:1]: I don't think anything bad happens if you call it multiple times, but yeah, this will be unnecessary after #323 merges.

File: packages/server/src/routes/tenants.js
Comment (ajhyndman) [inline packages/server/src/routes/tenants.js:34]: Are these routes used anywhere yet?  I'd just as soon not support creation of a tenant via API if not.

File: packages/server/src/db/tenant_creation.js
Comment (ajhyndman) [inline packages/server/src/db/tenant_creation.js:179]: This is definitely awkward.  And it should be redundant with the `main_agency_id` field in the `tenants` table, is that correct?

I would definitely prefer to land a migration that removes this redundancy and updates any queries rather than find a weird workaround for this self-reference.

**Review by ajhyndman — COMMENTED** [review]
I did give this a read through, and was able to run the script and verify it works!

I'm not as familiar with all the context here, so I left a couple questions.

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [reply packages/server/src/routes/tenants.js:34]: The "list tenants" and "update display name" do have UI that calls them; the "create tenant" one does not yet.

It's not pressing since we have the script implemented in #193, but this PR was open already and I wanted to make sure that if/when it did merge it correctly handled the ARPA tables.

File: packages/server/src/db/tenant_creation.js
Comment (mattbroussard) [reply packages/server/src/db/tenant_creation.js:179]: I think you're right that these fields are redundant and we could get rid of the one on `agencies`, though a few callsites might need to do an extra db call to get the `tenants` row.

**Comment by mhuang-usdr** [comment]
Sorry, I just saw these questions. >_<

>   * However, it looks like in prod there are lots of non-USDR users as "admins" with tenantId=1. @mhuang-usdr is there a different condition we should be using for "super admin", or should we clean up this prod data so only actual USDR staff/volunteers are in the USDR tenant?

We will be cleaning the USDR up so that only USDR users are admins with tenantID=1 and shifting everyone else out to a Demo tenant. We don't necessarily need a super-admin role in that case, but if the work is already done don't undo it!

File: packages/server/src/db/tenant_creation.js
Comment (mattbroussard) [reply packages/server/src/db/tenant_creation.js:179]: Filed this as a separate issue, #357. Will not be addressing in this PR.

File: packages/server/src/lib/access-helpers.js
Comment (mattbroussard) [reply packages/server/src/lib/access-helpers.js]: Punting on this for now because it adds some additional complexity (though if we ever want more than 2 roles we may end up adding such complexity anyhow). Instead, I'm restricting this "super admin" permission further by email (must be `@usdigitalresponse.org`) and filed a follow-up task #359 to clean up non-USDR users in USDR tenant and relax the email requirement.

**Comment by mattbroussard** [comment]
I think this should be ready to merge unless you have other changes you'd like to see @ajhyndman. If not, can you mark the PR approved so I can merge?

I've filed the following follow-up tasks to be handled separately:
- #357
- #358
- #359

File: packages/server/src/routes/tenants.js
Comment (mattbroussard) [reply packages/server/src/routes/tenants.js:34]: Filed follow-up task #358 to implement such UI.

**Review by mattbroussard — APPROVED** [review]

**Pull request merged**
